### PR TITLE
[None][perf] Cute DSL GVR Top-K: short-row remove cluster sync in run_one_row

### PR DESCRIPTION
**Pull request by limin2021**
## Description

Summary

For GVR Top-K with cluster_size > 1, each cluster of cs CTAs cooperates on a single row by dividing the N-element row into N/cs slices. When a row is short enough that its entire N fits within one CTA's design slice (max_slice_len = ceil(buffer_max / cluster_size)), the cluster cooperation overhead (6-10 cluster barriers per secant iteration) exceeds the compute saved.

This PR adds a runtime short-row degrade path:

- Long row (N > max_slice_len): all cs CTAs cooperate as before, each scanning its N/cs slice.
- Short row (N <= max_slice_len): CTA 0 solo-scans the full row with do_cluster_sync=False; the other cs-1 CTAs fall through run_one_row without doing any work.

#### 1.  Seqlen distributions (`max_seq_len=128K`)

Stats pooled across BS=[1, 64] (the range where `cluster_size > 1` is selected;
at larger BS the heuristic falls back to `cluster_size=1` regardless of row length).

| pattern | p25   | median | p75    | max (across BS=[1,64]) | description |
|---------|-------|--------|--------|------------------------|-------------|
| chat    | 1K    | 1K     | 1.1K   | **1.3K – 4K**          | all rows well below any degrade threshold; degrade fires on ~100% of rows |
| mixed   | 1K    | 1.8K   | 3.4K   | **2.9K – 28K**         | mostly short; degrade fires on nearly all rows |
| longdoc | 13.4K | 25.6K  | 42.6K  | **31K – 128K**         | spread across buffer; mix of degrade and co-op rows |
| random  | 48.4K | 72K    | 100K   | **13K – 128K**         | long rows dominate; most rows above degrade threshold |
| uniform | 128K  | 128K   | 128K   | **128K**               | all rows at `max_seq_len`; degrade never fires |

#### 2. performance

N: max sequence length in model level

##### bf16 K=512

| seqlen pattern | N=64K      | N=128K     | N=256K     |
|----------------|-----------|-----------|-----------|
| chat           | **+29.2%** | **+33.6%** | **+31.1%** |
| mixed          | **+20.6%** | **+23.4%** | **+24.1%** |
| longdoc        | -1.5%      | -1.5%      | -5.6%      |
| random         | **+2.6%**  | -1.1%      | -0.9%      |
| uniform        | -1.2%      | -0.8%      | -1.6%      |

##### bf16 K=1024

| seqlen pattern | N=64K      | N=128K     | N=256K     |
|----------------|-----------|-----------|-----------|
| chat           | **+44.0%** | **+54.8%** | **+48.0%** |
| mixed          | **+40.2%** | **+37.6%** | **+36.3%** |
| longdoc        | **+2.7%**  | **+2.8%**  | -0.7%      |
| random         | **+6.8%**  | **+4.4%**  | ~+0.3%     |
| uniform        | **+2.6%**  | **+2.6%**  | -1.2%      |

##### fp32 K=2048

| seqlen pattern | N=64K      | N=128K     | N=256K     |
|----------------|-----------|-----------|-----------|
| chat           | **+28.2%** | **+22.3%** | **+8.3%**  |
| mixed          | **+5.0%**  | -3.3%      | **+14.4%** |
| longdoc        | -2.9%      | **+16.3%** | **+13.6%** |
| random         | **+4.0%**  | -2.6%      | ~+0.4%     |
| uniform        | -1.3%      | **+2.7%**  | -4.8%      |

## Test Coverage

` python -m pytest tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py -v`

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/top_k/gvr_topk_decode.py [hunk 1]
```diff
@@ -2168,10 +2370,33 @@ def __call__(
...
+        # (UT / bench / future production) that enables the path without
+        # the matching size check.
+        if cutlass.const_expr(self.enable_smem_cache):
+            max_slice_len = (input_data.shape[1] + cluster_size - 1) // cluster_size
+            assert max_slice_len <= self.smem_cache_elems, (
+                f"enable_smem_cache=True requires per-CTA slice_len "
+                f"({max_slice_len}) <= smem_cache_elems "
+                f"({self.smem_cache_elems}); raise smem_cache_elems, "
+                f"increase cluster_size, or disable enable_smem_cache."
+            )
```
[the remark this answers is not in the document]
Comment (limin2021) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/top_k/gvr_topk_decode.py]: Removed this assert.

File: tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py [hunk 2]
```diff
@@ -267,7 +285,238 @@ def test_cute_dsl_gvr_topk_decode(
...
+
+    n_long = int(counters[0].item())
+    if scenario == "all_long":
+        assert n_long == batch_size, f"expected {batch_size} long, got {n_long}"
+    elif scenario == "all_short":
+        assert n_long == 0, f"expected 0 long, got {n_long}"
+    elif scenario == "mixed_half":
+        assert n_long == batch_size - batch_size // 2
+
+    _tie_aware_check(out_indices, logits, seq_lens, top_k, next_n, compress_ratio=1)
```
[the remark this answers is not in the document]
Comment (limin2021) [reply tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py:584]: Thanks for catching this. Fixed: _make_inputs now accepts an optional seq_lens parameter; for mixed_half, seq_lens is constructed first (short=8K, long=128K) and passed in before argmax is computed. This ensures pre_idx[..., 0] is within min(seq_lens)=8K for all rows. Also renamed override → seq_lens_mode with values "uniform" / "half_short_half_long" for clarity.

**Review by longcheng-nv — COMMENTED** [review]
Reviewed at head `18ea324`. I went through the new dispatch structure in detail — overall the implementation is clean, and I verified the following hold:

- All 5 cluster-primitive sites (`cluster_arrive/wait`, `mapa`, `ld_shared_cluster`) are gated by `do_cluster_sync`, and the long-vs-short decision is uniform within a cluster (same `row_idx` → same `N`), so there is no partial-participation barrier mismatch / deadlock surface. In degrade mode the leader issues no DSMEM accesses, so early-exiting peers freeing their SMEM is safe.
- Skipping the `s_iscalars[5]` snapshot in degrade mode is correct: the only reader is the Phase-4 DSMEM gather (also gated), and Phase 3 recomputes the CTA-local total from `smem_ptcnt` into `s_iscalars[0]`.
- The 26 positional args at all three `_run_phases` call sites match the signature.
- The degenerate-threshold branch stays balanced in long-row mode (identical preIdx stats across CTAs → same branch → all reach the tail barrier).
- Early `griddepcontrol_launch_dependents` from non-leader CTAs is benign (dependent grid waits for all CTAs; the leader signals after its writes).
- The leader-only rewrite of the `N <= top_k` path and the test-side `pre_idx[..., 0]` invariant fix are both genuine fixes.

### Main comment — degrade threshold is tied to buffer width, not to the actual break-even

`max_slice_len = ceil(input_data.shape[1] / cs)` is a function of the *allocation* width, but whether solo beats co-op is a function of absolute `N`: solo wins iff `t_scan(N) − t_scan(N/cs) < t_barriers`. When the buffer is much larger than typical rows (e.g. allocated for 512K, serving 64–128K requests), `max_slice_len` can reach 128K and a row with `N_eff ≈ 120K` gets degraded to a single CTA scanning 120K elements — where 4 CTAs scanning 30K each plus a few barriers is very likely faster.

This also contradicts the break-even already encoded in this same directory: `gvr_topk_decode_load_balance.py` uses `long_threshold = 64K` with the comment "cs=4 is B200 break-even", i.e. rows above ~64K *benefit* from cluster co-op. The two heuristics disagree whenever `ceil(buffer/cs) > 64K`.

Two ways to close this (either is fine):
1. Cap the threshold: `degrade_threshold = min(max_slice_len, <empirical crossover, same origin as LB's long_threshold>)` — no measurement needed; or
2. Show nsys data for the "large buffer + moderate N" cell demonstrating the degrade doesn't regress there.

### Suggestions (non-blocking)

- **Perf data**: since this op is not yet wired into any production path (only the custom-op registration + UT reference it), I don't consider benchmark numbers blocking — but if you have them from local runs, 2–4 nsys points around the threshold (e.g. `N_eff` at `max_slice_len` ±25%, one production (dtype, K) cell) would make the win concrete and would settle the point above with data.
- **Removed `enable_smem_cache` launch assert**: the removal rationale (n_cols is `cute.sym_int` at trace time) is valid, but this was the only backstop against a silent SMEM overrun. The Python-side `__call__` does see the concrete `input_data.shape[1]`, so a one-line host-side check (`ceil(n_cols/cs) > smem_cache_elems → raise`) would restore the guard without waiting for the cache-key rework in the TODO.
- **Deterministic boundary coverage**: the degrade path is currently exercised only probabilistically via `varlen=True` uniform sampling (seeded, so reproducible — with cs=4 roughly a quarter of rows land below `max_slice_len`). A small parametrized case with explicitly constructed `seq_lens` pinning `N_eff = max_slice_len` / ±1 plus a mixed batch (degrade + co-op rows in one launch) would pin the boundary, similar to what `test_lb_main_branches` does for the LB threshold.

### Nits

- The degenerate-threshold identity-emit block is duplicated across the `cs==1` / `cs>1` branches in `_run_phases`; since `is_leader` is constant-true at cs=1, a single `if is_leader & (tidx == 0):` covers both (or hoist the emit body into a helper if the const_expr split is intentional for IR flatness).
- The three `_run_phases` call sites pass 26 bare positional args; keyword-passing at least `slice_start` / `slice_end` / `do_cluster_sync` would make future insertions less error-prone (Phase 2/3 already use the keyword style for `do_cluster_sync`).

**Comment by limin2021** [comment]
@longcheng-nv 

#### for main comment
The threshold max_slice_len = ceil(buffer_width / cs) is the exact boundary at which CTA 1+ begin to have any valid tokens to scan. The co-op path uses static slice assignment based on buffer layout:

CTA k scans [k * max_slice_len, (k+1) * max_slice_len) ∩ [0, N_eff)

For the example you describe (buffer=512K, cs=4 → max_slice_len=128K, N_eff=120K):
- CTA 0: [0, 128K) ∩ [0, 120K) = 120K tokens (all the work)
- CTA 1: [128K, 256K) ∩ [0, 120K) = empty
- CTA 2, 3: also empty

Co-op here is identical to solo execution but with extra mbarrier overhead, so degrade is strictly correct.

The 4×30K picture would require dynamic slicing by N_eff (each CTA gets N_eff/cs tokens) — which is exactly what the load-balance path does, and why long_threshold = 64K makes sense there. The non-LB path cannot do dynamic slicing because slice boundaries must be fixed to the buffer layout for aligned/vectorized access. The two thresholds measure fundamentally different things and there is no contradiction.

The benchmark data supports this: for chat and mixed distributions where actual N_eff ≪ buffer_width, the degrade gives 30–50% speedup (see tables in the MR description). The longdoc N=256K case — where N_eff genuinely approaches max_slice_len — shows no meaningful regression.


#### perf data
Added in the MR description

#### test
 Added test_shortrow_degrade_boundary


#### nit 1
 Thanks for the note. The two _run_phases(do_cluster_sync=False) calls are in separate cutlass.const_expr(cluster_size > 1) branches, which CuTe DSL eliminates at trace time. A compiled cs=1 kernel contains only the else call; a compiled cs>1 kernel contains only the short-row call. They're mutually exclusive specializations, not duplicates in the compiled output.

**Review by hyukn — APPROVED** [review]

**Review by longcheng-nv — APPROVED** [review]
Re-reviewed at head `14c15dc`. All cluster-primitive sites are correctly gated by `do_cluster_sync`, the long-vs-short decision is uniform within a cluster (no partial-participation barrier hazard), and the degrade-path leader does no DSMEM access to early-exiting peers. The boundary test (max_slice_len +/- 1 + mixed batch) covers the transition point well, and the two bundled fixes (leader-only degenerate write, test pre_idx argmax fix) are both valid. The only CI failure at this head (B300 unittest/_torch/executor) is unrelated to this PR. LGTM.

**Review by mingyangHao — APPROVED** [review]
LGTM
